### PR TITLE
[Coverage] Introduce "partial fold" on BranchRegion

### PR DESCRIPTION
**Pull request by chapuni**
Currently both True/False counts were folded. It lost the information, "It is True or False before folding." It prevented recalling branch counts in merging template instantiations.

In `llvm-cov`, a folded branch is shown as:

- `[True: n, Folded]`
- `[Folded, False n]`

In the case If `n` is zero, a branch is reported as "uncovered". This is distinguished from "folded" branch. When folded branches are merged, `Folded` may be dissolved.

In the coverage map, either `Counter` is `Zero`. Currently both were `Zero`.

Since "partial fold" has been introduced, either case in `switch` is omitted as `Folded`.

Each `case:` in `switch` is reported as `[True: n, Folded]`, since `False` count doesn't show meaningful value.

When `switch` doesn't have `default:`, `switch (Cond)` is reported as `[Folded, False: n]`, since `True` count was just the sum of `case`(s). `switch` with `default` can be considered as "the statement that doesn't have any `False`(s)".

### CODE REVIEW

File: llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h [hunk 1]
```diff
@@ -719,10 +720,10 @@ struct FunctionRecord {
         Region.Kind == CounterMappingRegion::MCDCBranchRegion) {
       CountedBranchRegions.emplace_back(Region, Count, FalseCount,
                                         HasSingleByteCoverage);
-      // If both counters are hard-coded to zero, then this region represents a
+      // If either counters is hard-coded to zero, then this region represents a
```
Comment (ornata) [inline llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h:723]: s/counters/counter

Comment (ornata) [inline llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h:723]: s/counters/counter

File: llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h [hunk 2]
```diff
@@ -719,10 +720,10 @@ struct FunctionRecord {
...
       // constant-folded branch.
-      if (Region.Count.isZero() && Region.FalseCount.isZero())
-        CountedBranchRegions.back().Folded = true;
+      CountedBranchRegions.back().TrueFolded = Region.Count.isZero();
```
Comment (ornata) [inline llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h:725]: rename Region.Count to Region.TrueCount?

File: llvm/tools/llvm-cov/CoverageExporterJson.cpp [hunk 3]
```diff
@@ -125,7 +125,7 @@ json::Array renderRegions(ArrayRef<coverage::CountedRegion> Regions) {
 json::Array renderBranchRegions(ArrayRef<coverage::CountedRegion> Regions) {
   json::Array RegionArray;
   for (const auto &Region : Regions)
-    if (!Region.Folded)
+    if (!Region.TrueFolded || !Region.FalseFolded)
```
Comment (ornata) [inline llvm/tools/llvm-cov/CoverageExporterJson.cpp:128]: replace with a helper function? e.g. Region.isFolded()

File: llvm/tools/llvm-cov/CoverageSummaryInfo.cpp [hunk 4]
```diff
@@ -19,18 +19,18 @@ using namespace coverage;
...
-    if (BR.ExecutionCount > 0)
-      ++CoveredBranches;
-    // "False" Condition Branches.
-    ++NumBranches;
-    if (BR.FalseExecutionCount > 0)
-      ++CoveredBranches;
+    if (!BR.TrueFolded) {
+      // "True" Condition Branches.
+      ++NumBranches;
+      if (BR.ExecutionCount > 0)
```
Comment (ornata) [inline llvm/tools/llvm-cov/CoverageSummaryInfo.cpp:25]: Rename ExecutionCount to TrueExecutionCount?

File: llvm/tools/llvm-cov/CoverageSummaryInfo.cpp [hunk 5]
```diff
@@ -19,18 +19,18 @@ using namespace coverage;
...
+        ++CoveredBranches;
+    }
+    if (!BR.FalseFolded) {
```
Comment (ornata) [inline llvm/tools/llvm-cov/CoverageSummaryInfo.cpp:28]: Can this be a lambda or static helper function?

**Review by chapuni — COMMENTED** [review]
Thanks for comments!

File: llvm/tools/llvm-cov/CoverageSummaryInfo.cpp (hunk 5 above)
Comment (chapuni) [reply llvm/tools/llvm-cov/CoverageSummaryInfo.cpp:28]: I think this is enough short with expanded impl.

**Comment by chapuni** [comment]
@hanickadot I didn't think there'd be anything to imagine.

That said, I will arrange examples later. I still wonder how we could show "partial fold" more clearly and make it more consistent.

**Comment by hanickadot** [comment]
I have no additional list of thing Jessica pointed out, LGTM

**Review by hanickadot — APPROVED** [review]

**Comment by evodius96** [comment]
I think this is a meaningful enhancement to branch coverage. I don't have much to add to what's been said.  LGTM. Thanks!

**Review by evodius96 — APPROVED** [review]

File: llvm/tools/llvm-cov/CoverageExporterJson.cpp (hunk 3 above)
Comment (chapuni) [reply llvm/tools/llvm-cov/CoverageExporterJson.cpp:128]: I guess we have to revisit how to show folding. Let me feed this to future tasks.

**Comment by Lambdaris** [comment]
This patch also helps to distinguish constants from rust, in which constant may have only one counter hard coded to `Zero`, thank you!

Besides, I found that with this change we should also count false counter id when [calculating max `CounterID`](https://github.com/llvm/llvm-project/blob/eb9af19fbf13b91c9df8d6dc8553277169976520/llvm/lib/ProfileData/Coverage/CoverageMapping.cpp#L610). Otherwise some cases would break with "unexpected" (actually expected!) counter id from profile data. I does so at [#94137](https://github.com/llvm/llvm-project/pull/94137/commits/755d8f253fc23b1ef1001095e0310d5f0fac07ab) and it would be also great if some one picked it and opened a smaller pr.

**Comment by chapuni** [comment]
@Lambdaris I have #112730 as well.

**Pull request merged**